File: src/weaponinventorymod/gui/WimGuiButtonSpec.java
package weaponinventorymod.gui;

import com.fs.starfarer.api.ui.Alignment;

import java.awt.Color;

final class WimGuiButtonSpec<A> {
    final float width;
    final String label;
    final Color textColor;
    final A action;
    final boolean enabled;
    final Alignment alignment;
    final WimGuiButtonColors colors;
    final Color borderColor;
    final String tooltip;

    WimGuiButtonSpec(float width,
                     String label,
                     Color textColor,
                     A action,
                     boolean enabled,
                     Alignment alignment,
                     WimGuiButtonColors colors,
                     Color borderColor,
                     String tooltip) {
        this.width = width;
        this.label = label;
        this.textColor = textColor;
        this.action = action;
        this.enabled = enabled;
        this.alignment = alignment;
        this.colors = colors;
        this.borderColor = borderColor;
        this.tooltip = tooltip;
    }

    static <A> WimGuiButtonSpec<A> sameColor(float width,
                                             String label,
                                             Color textColor,
                                             A action,
                                             boolean enabled,
                                             Alignment alignment,
                                             Color color,
                                             Color borderColor) {
        return dimmedInner(width, label, textColor, action, enabled, alignment, color, borderColor);
    }

    static <A> WimGuiButtonSpec<A> sameColor(float width,
                                             String label,
                                             Color textColor,
                                             A action,
                                             boolean enabled,
                                             Alignment alignment,
                                             Color color,
                                             Color borderColor,
                                             String tooltip) {
        return dimmedInner(width, label, textColor, action, enabled, alignment, color, borderColor, tooltip);
    }

    static <A> WimGuiButtonSpec<A> dimmedInner(float width,
                                               String label,
                                               Color textColor,
                                               A action,
                                               boolean enabled,
                                               Alignment alignment,
                                               Color color,
                                               Color borderColor) {
        return dimmedInner(width, label, textColor, action, enabled, alignment, color, borderColor, null);
    }

    static <A> WimGuiButtonSpec<A> dimmedInner(float width,
                                               String label,
                                               Color textColor,
                                               A action,
                                               boolean enabled,
                                               Alignment alignment,
                                               Color color,
                                               Color borderColor,
                                               String tooltip) {
        return new WimGuiButtonSpec<A>(
                width,
                label,
                textColor,
                action,
                enabled,
                alignment,
                WimGuiButtonColors.dimmedInner(color),
                borderColor,
                tooltip);
    }

    static <A> WimGuiButtonSpec<A> toggle(float width,
                                          String label,
                                          Color textColor,
                                          A action,
                                          Alignment alignment,
                                          Color color,
                                          Color borderColor) {
        return dimmedInner(width, label, textColor, action, true, alignment, color, borderColor);
    }

    static <A> WimGuiButtonSpec<A> semantic(float width,
                                            String label,
                                            A action,
                                            boolean enabled,
                                            Color enabledFill,
                                            Color borderColor) {
        return semantic(width, label, action, enabled, enabledFill, borderColor, null);
    }

    static <A> WimGuiButtonSpec<A> semantic(float width,
                                            String label,
                                            A action,
                                            boolean enabled,
                                            Color enabledFill,
                                            Color borderColor,
                                            String tooltip) {
        return sameColor(
                width,
                label,
                enabled ? WimGuiStyle.WHITE_TEXT : WimGuiStyle.DISABLED_TEXT,
                action,
                enabled,
                Alignment.MID,
                enabled ? enabledFill : WimGuiStyle.DISABLED_BACKGROUND,
                borderColor,
                tooltip);
    }
}

File: src/main/kotlin/weaponsprocurement/gui/WimGuiButtonSpec.kt
package weaponsprocurement.gui

import com.fs.starfarer.api.ui.Alignment
import com.fs.starfarer.api.ui.TooltipMakerAPI
import java.awt.Color

class WimGuiButtonSpec<A>(
    @JvmField val width: Float,
    @JvmField val label: String?,
    @JvmField val textColor: Color,
    @JvmField val action: A,
    @JvmField val enabled: Boolean,
    @JvmField val alignment: Alignment,
    @JvmField val colors: WimGuiButtonColors,
    @JvmField val borderColor: Color,
    @JvmField val tooltip: String?,
    @JvmField val tooltipCreator: TooltipMakerAPI.TooltipCreator? = null,
) {
    companion object {
        @JvmStatic
        fun <A> sameColor(
            width: Float,
            label: String?,
            textColor: Color,
            action: A,
            enabled: Boolean,
            alignment: Alignment,
            color: Color,
            borderColor: Color,
        ): WimGuiButtonSpec<A> = dimmedInner(width, label, textColor, action, enabled, alignment, color, borderColor)

        @JvmStatic
        fun <A> sameColor(
            width: Float,
            label: String?,
            textColor: Color,
            action: A,
            enabled: Boolean,
            alignment: Alignment,
            color: Color,
            borderColor: Color,
            tooltip: String?,
        ): WimGuiButtonSpec<A> = dimmedInner(width, label, textColor, action, enabled, alignment, color, borderColor, tooltip)

        @JvmStatic
        fun <A> dimmedInner(
            width: Float,
            label: String?,
            textColor: Color,
            action: A,
            enabled: Boolean,
            alignment: Alignment,
            color: Color,
            borderColor: Color,
        ): WimGuiButtonSpec<A> = dimmedInner(width, label, textColor, action, enabled, alignment, color, borderColor, null)

        @JvmStatic
        fun <A> dimmedInner(
            width: Float,
            label: String?,
            textColor: Color,
            action: A,
            enabled: Boolean,
            alignment: Alignment,
            color: Color,
            borderColor: Color,
            tooltip: String?,
        ): WimGuiButtonSpec<A> = WimGuiButtonSpec(
            width,
            label,
            textColor,
            action,
            enabled,
            alignment,
            WimGuiButtonColors.dimmedInner(color),
            borderColor,
            tooltip,
        )

        @JvmStatic
        fun <A> toggle(
            width: Float,
            label: String?,
            textColor: Color,
            action: A,
            alignment: Alignment,
            color: Color,
            borderColor: Color,
        ): WimGuiButtonSpec<A> = toggle(width, label, textColor, action, alignment, color, borderColor, null)

        @JvmStatic
        fun <A> toggle(
            width: Float,
            label: String?,
            textColor: Color,
            action: A,
            alignment: Alignment,
            color: Color,
            borderColor: Color,
            tooltip: String?,
        ): WimGuiButtonSpec<A> = dimmedInner(width, label, textColor, action, true, alignment, color, borderColor, tooltip)

        @JvmStatic
        fun <A> toggle(
            width: Float,
            label: String?,
            textColor: Color,
            action: A,
            alignment: Alignment,
            color: Color,
            borderColor: Color,
            tooltip: String?,
            tooltipCreator: TooltipMakerAPI.TooltipCreator?,
        ): WimGuiButtonSpec<A> = WimGuiButtonSpec(
            width,
            label,
            textColor,
            action,
            true,
            alignment,
            WimGuiButtonColors.dimmedInner(color),
            borderColor,
            tooltip,
            tooltipCreator,
        )

        @JvmStatic
        fun <A> semantic(
            width: Float,
            label: String?,
            action: A,
            enabled: Boolean,
            enabledFill: Color,
            borderColor: Color,
        ): WimGuiButtonSpec<A> = semantic(width, label, action, enabled, enabledFill, borderColor, null)

        @JvmStatic
        fun <A> semantic(
            width: Float,
            label: String?,
            action: A,
            enabled: Boolean,
            enabledFill: Color,
            borderColor: Color,
            tooltip: String?,
        ): WimGuiButtonSpec<A> = sameColor(
            width,
            label,
            if (enabled) WimGuiStyle.WHITE_TEXT else WimGuiStyle.DISABLED_TEXT,
            action,
            enabled,
            Alignment.MID,
            if (enabled) enabledFill else WimGuiStyle.DISABLED_BACKGROUND,
            borderColor,
            tooltip,
        )
    }
}
